### PR TITLE
Date range filters should respect the start-of-week Setting 📆 

### LINKED ISSUE #13604: Wrong day names are displayed when using not-Sunday as start of the week and grouping by "Day of week"
When grouping by day of the week, the DATA use the correct date but the AXIS LABELS still start with Sunday.

**To Reproduce**
- Start of the week: Monday
- Sample dataset: table Orders
- Filter: "Created at" between 2020-03-02 and 2020-03-03 (monday and tuesday)
- Summarize: by Count and group by Created At (by day of the week)
- Visualization: Bar, X-axis: Ordinal

**Expected behavior**
- Two bars, one for Monday 2020-03-02 labelled "Monday", one for Tuesday 2020-03-03 labelled "Tuesday"

**Actual behavior**
- Two bars, one for Monday 2020-03-02 labelled "Sunday", one for Tuesday 2020-03-03 labelled "Monday"

Capture d’écran de 2020-10-27 14-27-02

**Information about your Metabase Installation:**
- Metabase version: 0.37.0.2
- Metabase hosting environment: Jar-file on Ubuntu 16.04
- Metabase internal database: H2

**Severity**
Rather annoying. We do a lot of week-over-week stats (data from our helpdesk). Until now we had to make do with graphs starting with Sunday, which look lopsided (not a lot of issues opened on a Sunday, so the graphs look like calm Mondays and busy Tuesdays). This bug makes the graphs look correct (busy on week start, calm on weekends) but can't be used with wrong labels.

### LINKED ISSUE #14294: Start of Week not applied to Field Filter in Native question, which can lead to incorrect results
**Describe the bug**
Regardless of which day "start of week" is configured to, using the Field Filter in a Native Query will return results with Sunday as the start of the week.

**To Reproduce**

This bug is in current build, v1.37.3
1. Go to Settings > Admin > Localization
2. Change "First Day of Week" to Monday
3. Click "Ask A Question" > "Native Query" 
4. Select the Sample Dataset
5. Enter the following query
```
select ID, CREATED_AT, dayname(CREATED_AT) as CREATED_AT_DAY
from ORDERS 
[[where {{date_range}}]]
order by CREATED_AT
```
6. Configure the Variables Settings as follows:
Variable type: Field Filter
Field to map to: Order > Created At
Filter widget type: Date Filter
7. Set the filter to include the previous 52 weeks

**Expected behavior**
Since this sample dataset has a record for every day between Jan 1, 2020 - April 19, 2020 the oldest record for this filter should match the day of week we specified in "Start of Week".  

**Observed Behavior**
Regardless of what I configure the "Start of Week" day to, the filter always considers Sunday the start of the week.

**Screenshots**
startsatsunday

At the time of running this query, the start of "Previous 52 Weeks" should be Monday, Jan 6 2020.  However, the query results include Sunday, Jan 5th.  

**Information about your Metabase Installation:**
v1.37.3, docker running locally

**Severity**
This affects weekly reporting for customers that have their start of week set to something other than Sunday.  

Redshift treats Monday as the first day of the week in DATE_TRUNC, so this incongruity causes visualizations to have an extra week with values incorrectly offset in groupings.

### LINKED ISSUE #15044: Oracle, BigQuery filtering by column with day-of-week bucketing not working
Couldn't find any existing matching bug reports. While working on a fix for #13604 I noticed tests were failing for Oracle and BigQuery because we were generating queries that used `%` for modulus where BigQuery and Oracle instead use a [`mod()`](https://docs.oracle.com/cd/B19306_01/server.102/b14200/functions088.htm) function (actually  BigQuery can apparently use `%` *outside* of the `WHERE` clause, but has to use `mod()` in the `WHERE` clause). 

The BigQuery error was

```
Syntax error: Parenthesized expression cannot be parsed as an expression, struct constructor, or subquery
```

The Oracle error was

```
java.sql.SQLSyntaxErrorException: ORA-00920: invalid relational operator
```

### PR DESCRIPTION
**Pull request by camsaul**
Fixes #14294
Fixes #13604
Fixes #15044

`:day-of-week`  and `:week-of-year` now use the `start-of-week` Setting. I removed `:iso-day-of-week` and `:iso-week` from the date utils since they weren't actually used anywhere and you can just use the other units now instead. 

While working on this I realized #15039 is something we need to address in the future.

### CODE REVIEW

**Comment by camsaul** [comment]
Fixed for #13604

image

File: src/metabase/util/date_2.clj [hunk 1]
```diff
@@ -241,7 +243,7 @@
      :hour        (t/truncate-to t :hours)
      :day         (t/truncate-to t :days)
      :week        (-> (.with t (adjuster :first-day-of-week))     (t/truncate-to :days))
-     :iso-week    (-> (.with t (adjuster :first-day-of-iso-week)) (t/truncate-to :days))
+     ;; TODO -- we should remove this now that `:week` respects the `start-of-week` Setting.
```
Comment (camsaul) [inline src/metabase/util/date_2.clj:246]: oops. I did remove `:iso-week`, forgot to remove the `TODO`

File: test/metabase/util/date_2_test.clj [hunk 2]
```diff
@@ -378,6 +398,22 @@
...
 
+(deftest comparison-range-start-of-week-test
+  (testing "`comparison-range` for week should respect the `start-of-week` Setting (#14294)"
+    (doseq [[first-day-of-week expected] {"sunday" {:start #t "2021-02-21", :end #t "2021-02-27"}
+                                          "monday" {:start #t "2021-02-22", :end #t "2021-02-28"}
+                                          #_       "tuesday"   #_ [9 "Tuesday"]
+                                          #_       "wednesday" #_ [10 "Wednesday"]
+                                          #_       "thursday"  #_ [11 "Thursday"]
+                                          #_       "friday"    #_ [12 "Friday"]
+                                          #_       "saturday"  #_ [13 "Saturday"]}]
```
Comment (camsaul) [inline test/metabase/util/date_2_test.clj:409]: forgot to update/uncomment these. I'll fix in a second

File: src/metabase/util/date_2.clj [hunk 3]
```diff
@@ -121,24 +122,33 @@
...
-;; see https://clojure.atlassian.net/browse/TANAL-132
-(defn- week-fields ^WeekFields [k]
-  (get week-fields* k))
+(def ^:private ^{:arglists '(^java.time.DayOfWeek [k])} day-of-week*
+  (common/static-instances DayOfWeek))
+
+(defn- week-fields
+  "Create a new instance of a `WeekFields`, which is used for localized day-of-week, week-of-month, and week-of-year.
+
+    (week-fields :monday) ; -> #object[java.time.temporal.WeekFields \"WeekFields[TUESDAY,1]\"]"
```
Comment (camsaul) [inline src/metabase/util/date_2.clj:143]: this example is wrong

File: src/metabase/driver/sql/query_processor.clj [hunk 4]
```diff
@@ -116,15 +116,22 @@
                                   (- offset) :day)
       (truncate-fn expr))))
 
-(defn adjust-day-of-week
+(s/defn adjust-day-of-week
```
Comment (dpsutton) [inline src/metabase/driver/sql/query_processor.clj:119]: just curious, is there a benefit to schema's defn when you provide no type information on return or parameters?

Comment (dpsutton) [reply src/metabase/driver/sql/query_processor.clj:119]: nevermind i missed it was annotated in the second arity

File: test/metabase/util/date_2_test.clj [hunk 5]
```diff
@@ -250,7 +261,6 @@
             :hour     (t/zoned-date-time  2019 10 27 14 0  0 0 (t/zone-id "America/Los_Angeles"))
             :day      (t/zoned-date-time  2019 10 27  0 0  0 0 (t/zone-id "America/Los_Angeles"))
             :week     (t/zoned-date-time  2019 10 27  0 0  0 0 (t/zone-id "America/Los_Angeles"))
-            :iso-week (t/zoned-date-time  2019 10 21  0 0  0 0 (t/zone-id "America/Los_Angeles"))
```
Comment (dpsutton) [inline test/metabase/util/date_2_test.clj:253]: i'm not sure if i know enough for this question to make sense, but is it possible any cards or queries or questions or what-have-yous have `:iso-week` in them such that we need to support the offset still?

**Review by dpsutton — APPROVED** [review]

File: test/metabase/util/date_2_test.clj (hunk 5 above)
Comment (camsaul) [reply test/metabase/util/date_2_test.clj:253]: We've never used `:iso-week` or the other ones I removed anywhere outside of the date util namespace. When I wrote that date util namespace we only supported Sunday-start weeks, so I put those `:iso-` units were in there mostly to make implementing ISO week support easier when we actually got around to it. We didn't end up implementing ISO week support -- we implemented the more general `start-of-week` Setting, so `:iso-week` and the like never ended up being used anywhere.

Anyways the valid units in Cards/queries/etc. are all defined in `metabase.mbql.schema` and regardless of what's in here you can't use anything not in the MBQL schema. The `:iso-` options were never in there

**Comment by nemanjaglumac** [comment]
There was a repro for #14924. I unskipped it in #15087.

**Pull request merged**